MekHQ/src/mekhq/campaign/mission/CrewSkillUpgrader.java

```diff
@@ -216,7 +221,16 @@ private boolean extraEligibilityCheck(SpecialAbility spa, Entity entity) {
      * @param entity The entity being manipulated
      * @return Weapon name
      */
-    private String pickRandomWeapon(Entity entity) {
+    private String pickRandomWeapon(Entity entity, boolean clusterOnly) {
+        List<Mounted> weapons = entity.getIndividualWeaponList();
+        List<Mounted> eligibleWeapons = new ArrayList<>();
+        
+        for(Mounted weapon : weapons) {
+            if(SpecialAbility.isWeaponEligibleForSPA(weapon.getType(), Person.T_NONE, clusterOnly)) {
+                eligibleWeapons.add(weapon);
+            }
+        }
+        
         int weaponIndex = Compute.randomInt(entity.getIndividualWeaponList().size());
         return entity.getIndividualWeaponList().get(weaponIndex).getName();
```

Comment:
After building a list of all eligible weapons, you're still picking randomly from all weapons.